core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java

```diff
@@ -258,20 +303,53 @@ private static long optionalSum(long left, long right) {
             }
             mergedTopDocs = TopDocs.merge(from, topN, shardTopDocs);
         }
-        return mergedTopDocs.scoreDocs;
+
+        final Map<String, List<Suggestion>> groupedCompletionSuggestions = new TreeMap<>();
+        // group suggestions and assign shard index
+        for (AtomicArray.Entry<? extends QuerySearchResultProvider> sortedResult : sortedResults) {
+            if (sortedResult.value.queryResult().suggest() != null) {
+                for (Suggestion<?> suggestion : sortedResult.value.queryResult().suggest()) {
+                    if (suggestion instanceof CompletionSuggestion) {
+                        suggestion.setShardIndex(sortedResult.index);
+                        List<Suggestion> suggestions =
+                            groupedCompletionSuggestions.computeIfAbsent(suggestion.getName(), s -> new ArrayList<>());
+                        suggestions.add(suggestion);
+                    }
+                }
+            }
+        }
+
+        ScoreDoc[] scoreDocs = mergedTopDocs.scoreDocs;
+        final int[] offsets = new int[1 + groupedCompletionSuggestions.size()];
+        int offsetIndex = 0;
+        offsets[offsetIndex] = 0;
+        for (List<Suggestion> completionSuggestions : groupedCompletionSuggestions.values()) {
+            final CompletionSuggestion completionSuggestion = CompletionSuggestion.reduceTo(completionSuggestions);
+            assert completionSuggestion != null;
+            offsets[++offsetIndex] = scoreDocs.length;
+            List<CompletionSuggestion.Entry.Option> options = completionSuggestion.getEntries().get(0).getOptions();
+            ScoreDoc[] newScoreDocs = new ScoreDoc[scoreDocs.length + options.size()];
+            System.arraycopy(scoreDocs, 0, newScoreDocs, 0, scoreDocs.length);
+            for (int i = 0; i < options.size(); i++) {
+                newScoreDocs[scoreDocs.length + i] = options.get(i).getDoc();
+            }
+            scoreDocs = newScoreDocs;
+        }
+        return new SortedDocs(scoreDocs, offsets);
     }
 
-    public ScoreDoc[] getLastEmittedDocPerShard(SearchRequest request, ScoreDoc[] sortedShardList, int numShards) {
+    public ScoreDoc[] getLastEmittedDocPerShard(SearchRequest request, SortedDocs sortedShardList, int numShards) {
         if (request.scroll() != null) {
             return getLastEmittedDocPerShard(sortedShardList, numShards);
         } else {
             return null;
         }
     }
 
-    public ScoreDoc[] getLastEmittedDocPerShard(ScoreDoc[] sortedShardList, int numShards) {
+    public ScoreDoc[] getLastEmittedDocPerShard(SortedDocs sortedDocs, int numShards) {
         ScoreDoc[] lastEmittedDocPerShard = new ScoreDoc[numShards];
-        for (ScoreDoc scoreDoc : sortedShardList) {
+        for (int sortedDocsIndex = sortedDocs.getOffset(0); sortedDocsIndex < sortedDocs.getOffset(1); sortedDocsIndex++) {
```

Comment:
if so, I think we should have a signature like: `public ScoreDoc[] getLastEmittedDocPerShard(ScoreDocs[] sortedShardList, int numShards, int offset, int length) {` it's more flexible I think